src/box/gc.c

```diff
@@ -744,6 +822,411 @@ gc_consumer_update(struct gc_consumer *consumer, const struct vclock *vclock)
+		opts = tuple_field_with_type(tuple, BOX_GC_CONSUMERS_FIELD_OPTS,
+					     MP_MAP);
+	}
+	if (gc_consumer_def_opts_decode(def, opts, region) != 0)
+		return NULL;
+	return def;
+}
+
+/**
+ * Data passed to transactional triggers of replace in _gc_consumers.
+ */
+struct gc_consumers_txn_trigger_data {
+	/*
+	 * Saved old definition of consumer.
+	 */
+	struct gc_consumer_def *old_def;
+	/*
+	 * Saved new definition of consumer.
+	 */
+	struct gc_consumer_def *new_def;
+};
+
+static int
+on_replace_dd_gc_consumers_commit(struct trigger *trigger, void *event)
+{
+	(void)event;
+	struct gc_consumers_txn_trigger_data *data =
+		(struct gc_consumers_txn_trigger_data *)trigger->data;
+	struct gc_consumer_def *old_def = data->old_def;
+	struct gc_consumer_def *new_def = data->new_def;
+
+	/* Unref GC delay on first consumer. */
+	if (new_def != NULL && old_def == NULL)
+		gc_delay_unref();
+
+	if (old_def == NULL) {
+		/*
+		 * INSERT
+		 * Despite the consumer just appeared, it can be already
+		 * registered. It can happen on schema upgrade - all replicas
+		 * had in-memory consumers until the space _gc_consumers is
+		 * created.
+		 */
+		struct gc_consumer *consumer =
+			gc_consumer_by_uuid(&new_def->uuid);
+		if (consumer == NULL) {
+			consumer = gc_consumer_register_impl(
+				&new_def->vclock, new_def->object,
+				&new_def->uuid);
+			if (consumer == NULL)
+				return -1;
+		} else {
+			gc_consumer_update_impl(consumer, &new_def->vclock);
+		}
+		consumer->is_persistent = true;
+	} else if (new_def == NULL) {
+		/* DELETE */
+		struct gc_consumer *consumer =
+			gc_consumer_by_uuid(&old_def->uuid);
+		assert(consumer != NULL);
+		consumer->is_persistent = false;
+		if (consumer->ref_count == 0) {
+			gc_consumer_unregister_impl(consumer);
+		} else if (!consumer->is_inactive) {
+			/* Deactivate consumer if it is still used in-memory. */
+			consumer->is_inactive = true;
+			gc_tree_remove(&gc.active_consumers, consumer);
+		}
+	} else {
+		/* UPDATE */
+		struct gc_consumer *consumer =
+			gc_consumer_by_uuid(&old_def->uuid);
+		assert(consumer != NULL);
+		/* Update the name to be consistent with persistent state. */
+		strlcpy(consumer->object_name, new_def->object, GC_NAME_MAX);
+		gc_consumer_update_impl(consumer, &new_def->vclock);
+	}
+	return 0;
+}
+
+/**
+ * Retry updating persistent WAL GC state every N seconds.
+ * Retry timeout should be quite large since there is no point
+ * to write to database too often if we fail to do it. In debug
+ * mode retry more often for testing purposes.
+ */
+#ifdef NDEBUG
+enum { WAL_GC_PERSIST_FIBER_RETRY_TIMEOUT = 10 };
+#else
+enum { WAL_GC_PERSIST_FIBER_RETRY_TIMEOUT = 1 };
+#endif
+
+static int
+gc_consumers_persist_fiber_f(va_list ap)
+{
+	(void)ap;
+	const double retry_timeout = WAL_GC_PERSIST_FIBER_RETRY_TIMEOUT;
+	bool say_once = false;
+	while (!fiber_is_cancelled()) {
+		fiber_check_gc();
+
+		int64_t delta = gc.persist_scheduled - gc.persist_completed;
+		if (delta == 0) {
```

Comment:
I think that the fiber should be rate-limited, otherwise it may consumer quite a lot of cpu/disk. I think it's totally fine to sync the state once every 10 seconds or so.

Reply:
The fiber is woken up extremely rare (when clog is closed by relay), so it's not needed to be rate-limited.